### PR TITLE
Allow blobs to have dependancies.

### PR DESCRIPTION
**Pull request by mikalstill**
This is needed to for docker images, which are naturally a series of layers, but also makes sense for VM images when we want thin snapshots.

### CODE REVIEW

**Comment by mikalstill** [comment]
This is still a work in progress, I need to add the API calls to create a thin snapshot, ensure that a VM can be constructed with a non-trivial list of dependant blobs (I suspect I need to add more backing file entries to the libvirt template), and needs some solid CI testing.

**Comment by mikalstill** [comment]
@mandoonandy, I think this is ready for review as a logical boundary on my adventures. I propose this would stay in develop until 0.6, so landing this implies that other things we want in 0.5 releases would need to be backported.

**Review by mandoonandy — CHANGES_REQUESTED** [review]
Looks good, minor questions

File: shakenfist/blob.py [hunk 1]
```diff
@@ -345,19 +385,31 @@ def ensure_blob_path():
...
     # Actually make the snapshot
+    depends_on = None
     with util_general.RecordedOperation('snapshot %s' % disk['device'], related_object):
-        util_image.snapshot(None, disk['path'], dest_path)
+        depends_on = util_image.snapshot(
+            None, disk['path'], dest_path, thin=thin)
         st = os.stat(dest_path)
 
+    # Check that the dependancy (if any) actually exists. This test can fail when
+    # the blob used to start an instance has been deleted already.
```
[the remark this answers is not in the document]
Comment (mikalstill) [reply shakenfist/blob.py:402]: I kind of meant this comment as a hedge. We do reference count the dependancies, so this should not occur except in an exceptional case.

File: shakenfist/blob.py [hunk 2]
```diff
@@ -15,7 +15,7 @@
 from shakenfist.constants import LOCK_REFRESH_SECONDS, GiB
 from shakenfist import db
 from shakenfist import etcd
-from shakenfist.exceptions import BlobDeleted, BlobFetchFailed
+from shakenfist.exceptions import BlobDeleted, BlobFetchFailed, BlobDependancyMissing
```
[the remark this answers is not in the document]
Comment (mikalstill) [reply shakenfist/blob.py:18]: Done.

File: shakenfist/blob.py [hunk 3]
```diff
@@ -45,16 +45,21 @@ def __init__(self, static_values):
         self.__size = static_values['size']
         self.__modified = static_values['modified']
         self.__fetched_at = static_values['fetched_at']
+        self.__depends_on = static_values['depends_on']
+        self.__is_transcode_of = static_values['is_transcode_of']
 
     @classmethod
-    def new(cls, blob_uuid, size, modified, fetched_at):
+    def new(cls, blob_uuid, size, modified, fetched_at, depends_on=None,
+            is_transcode_of=None):
```
[the remark this answers is not in the document]
Comment (mikalstill) [reply shakenfist/blob.py:53]: Done, but I decided I liked `transcodes`.

File: shakenfist/client/upgrade.py [hunk 4]
```diff
@@ -110,6 +110,23 @@ def main():
                     etcd_client.put(metadata['key'],
                                     json.dumps(i, indent=4, sort_keys=True))
 
+        if minor <= 5:
+            for data, metadata in etcd_client.get_prefix('/sf/blob/'):
+                b = json.loads(data.decode('utf-8'))
+                changed = False
+
+                # Find version 3 instances and migrate them to version 4
+                if b.get('version') == 2:
```
[the remark this answers is not in the document]
Comment (mikalstill) [reply shakenfist/client/upgrade.py:119]: I think this is a more general question we should deal with outside this branch -- when do we start supporting online upgrades? How do we support them? What do we do with old versions of SF that are running during the upgrade and need the old version?

File: shakenfist/images.py [hunk 5]
```diff
@@ -103,63 +118,97 @@ def transcode_image(self, lock, b):
...
+                        [lock], 'gunzip -k -q -c %s > %s' % (blob_path, cache_path))
+                blob_path = cache_path
+
+            cache_path = os.path.join(
+                config.STORAGE_PATH, 'image_cache', b.uuid + '.qcow2')
+            cache_info = util_image.identify(blob_path)
+
+            # Convert the cluster size from qemu format to an int
+            cluster_size_as_int = QCOW2_CLUSTER_SIZE
+            if cluster_size_as_int.endswith('M'):
```
[the remark this answers is not in the document]
Comment (mikalstill) [reply shakenfist/images.py:173]: Done.

File: shakenfist/blob.py [hunk 6]
```diff
@@ -345,19 +385,31 @@ def ensure_blob_path():
...
 
     # Actually make the snapshot
+    depends_on = None
     with util_general.RecordedOperation('snapshot %s' % disk['device'], related_object):
-        util_image.snapshot(None, disk['path'], dest_path)
+        depends_on = util_image.snapshot(
+            None, disk['path'], dest_path, thin=thin)
         st = os.stat(dest_path)
 
+    # Check that the dependancy (if any) actually exists. This test can fail when
```
Comment (mandoonandy) [inline shakenfist/blob.py:401]: dependency

**Review by mandoonandy — APPROVED** [review]
LGTM

**Pull request merged**